File: backend/src/main/java/com/factorycare/backend/domain/fault/entity/Fault.java
package com.factorycare.backend.domain.fault.entity;

import com.factorycare.backend.domain.equipment.entity.Equipment;
import com.factorycare.backend.domain.inspection.entity.InspectionResult;
import com.factorycare.backend.domain.user.entity.User;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "faults")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public class Fault {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "equipment_id", nullable = false)
    private Equipment equipment;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FaultSeverity severity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FaultStatus status = FaultStatus.REPORTED;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reported_by_id", nullable = false)
    private User reportedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_id")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "inspection_result_id")
    private InspectionResult inspectionResult;

    @OneToMany(mappedBy = "fault", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("changedAt ASC")
    private List<FaultStatusHistory> statusHistories = new ArrayList<>();

    private LocalDateTime resolvedAt;

    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @Builder
    public Fault(Equipment equipment, String title, String description,
                 FaultSeverity severity, User reportedBy, InspectionResult inspectionResult) {
        this.equipment = equipment;
        this.title = title;
        this.description = description;
        this.severity = severity != null ? severity : FaultSeverity.MEDIUM;
        this.reportedBy = reportedBy;
        this.inspectionResult = inspectionResult;
    }

    public void update(String title, String description, FaultSeverity severity) {
        if (title != null) this.title = title;
        if (description != null) this.description = description;
        if (severity != null) this.severity = severity;
    }

    public void changeStatus(FaultStatus newStatus) {
        validateTransition(this.status, newStatus);
        if (newStatus == FaultStatus.RESOLVED) {
            this.resolvedAt = LocalDateTime.now();
        }
        this.status = newStatus;
    }

    public void assignTo(User user) {
        this.assignedTo = user;
    }

    private void validateTransition(FaultStatus from, FaultStatus to) {
        boolean valid = switch (from) {
            case REPORTED -> to == FaultStatus.CONFIRMED;
            case CONFIRMED -> to == FaultStatus.IN_PROGRESS;
            case IN_PROGRESS -> to == FaultStatus.RESOLVED;
            case RESOLVED -> to == FaultStatus.CLOSED || to == FaultStatus.IN_PROGRESS;
            case CLOSED -> false;
        };
        if (!valid) throw new IllegalStateException(from + " → " + to + " 상태 전이 불가");
    }
}

File: frontend/src/types/fault.ts
import type { SpringPage } from './equipment'

export type { SpringPage }

export type FaultSeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
export type FaultStatus = 'REPORTED' | 'CONFIRMED' | 'IN_PROGRESS' | 'RESOLVED' | 'CLOSED'

export const FAULT_SEVERITY_LABELS: Record<FaultSeverity, string> = {
  LOW: '낮음',
  MEDIUM: '중간',
  HIGH: '높음',
  CRITICAL: '심각',
}

export const FAULT_SEVERITY_COLORS: Record<FaultSeverity, string> = {
  LOW: 'bg-gray-100 text-gray-800',
  MEDIUM: 'bg-yellow-100 text-yellow-800',
  HIGH: 'bg-orange-100 text-orange-800',
  CRITICAL: 'bg-red-100 text-red-800',
}

export const FAULT_STATUS_LABELS: Record<FaultStatus, string> = {
  REPORTED: '접수',
  CONFIRMED: '확인',
  IN_PROGRESS: '작업중',
  RESOLVED: '해결',
  CLOSED: '완료',
}

export const FAULT_STATUS_COLORS: Record<FaultStatus, string> = {
  REPORTED: 'bg-blue-100 text-blue-800',
  CONFIRMED: 'bg-purple-100 text-purple-800',
  IN_PROGRESS: 'bg-yellow-100 text-yellow-800',
  RESOLVED: 'bg-green-100 text-green-800',
  CLOSED: 'bg-gray-100 text-gray-800',
}

export const NEXT_STATUS_MAP: Partial<Record<FaultStatus, FaultStatus[]>> = {
  REPORTED: ['CONFIRMED'],
  CONFIRMED: ['IN_PROGRESS'],
  IN_PROGRESS: ['RESOLVED'],
  RESOLVED: ['CLOSED', 'IN_PROGRESS'],
}

export interface FaultStatusHistory {
  id: number
  fromStatus: FaultStatus
  toStatus: FaultStatus
  changedByName: string
  reason: string | null
  changedAt: string
}

export interface Fault {
  id: number
  equipmentId: number
  equipmentName: string
  title: string
  description: string | null
  severity: FaultSeverity
  status: FaultStatus
  reportedById: number
  reportedByName: string
  assignedToId: number | null
  assignedToName: string | null
  inspectionResultId: number | null
  resolvedAt: string | null
  statusHistories: FaultStatusHistory[]
  createdAt: string
}

export interface FaultCreateRequest {
  equipmentId: number
  title: string
  description?: string
  severity?: FaultSeverity
}

export interface FaultUpdateRequest {
  title?: string
  description?: string
  severity?: FaultSeverity
}

export interface FaultStatusChangeRequest {
  status: FaultStatus
  reason?: string
}

export interface FaultAssignRequest {
  assigneeId: number
}

export interface FaultSearchParams {
  equipmentId?: number
  status?: FaultStatus
  severity?: FaultSeverity
  assigneeId?: number
  from?: string
  to?: string
  page?: number
  size?: number
}
